crates/circuit/src/register.rs

```diff
@@ -0,0 +1,1054 @@
+                // When creating `Owning` register, we don't need to create the `BitInfo`
+                // instances, they can be entirely derived from `self`.
+                Self(
+                    RegisterInfo::new_owning(Self::name_parse(name), size, Some($extra_exp)).into(),
+                )
+            }
+
+            /// Creates a Register whose bits already exist.
+            pub fn new_alias(name: Option<String>, bits: IndexSet<$bit>) -> Self {
+                Self(
+                    RegisterInfo::new_alias(
+                        Self::name_parse(name),
+                        Box::new(bits.into_iter().map(|bit| bit.0).collect()),
+                        Some($extra_exp),
+                    )
+                    .into(),
+                )
+            }
+        }
+
+        impl $name {
+            /// Provide a counted reference to the inner data of the register
+            pub(crate) fn data(&self) -> &Arc<RegisterInfo> {
+                &self.0
+            }
+
+            pub(crate) fn get_instance_count() -> u32 {
+                $counter.load(std::sync::atomic::Ordering::Relaxed)
+            }
+        }
+
+        impl Register for $name {
+            type Bit = $bit;
+            /// A reference to the register's name
+            fn name(&self) -> &str {
+                self.0.name()
+            }
+
+            /// Returns the size of the register.
+            fn len(&self) -> usize {
+                self.0.len()
+            }
+
+            /// Returns whether the register is empty.
+            fn is_empty(&self) -> bool {
+                self.0.is_empty()
+            }
+
+            /// Returns an iterator over the bits within the circuit
+            fn bits(&self) -> impl ExactSizeIterator<Item = <Self as Register>::Bit> {
+                self.0.bits().map(|bit| $bit(bit))
+            }
+
+            /// Checks if a bit is contained within the register
+            fn contains(&self, bit: &<$name as Register>::Bit) -> bool {
+                self.0.contains(&bit.0)
+            }
+
+            /// Gets a bit via index, return None if not present
+            fn get(&self, index: usize) -> Option<$bit> {
+                self.0.get(index).map(|bit| $bit(bit))
+            }
+        }
+    };
+}
+
+create_register_object! {QuantumRegister, ShareableQubit, BitExtraInfo::Qubit{is_ancilla: false}, QREG_COUNTER, "q"}
+impl QuantumRegister {
+    /// Check if the [QuantumRegister] instance is ancillary.
+    pub fn is_ancilla(&self) -> bool {
+        match self.0.as_ref() {
+            RegisterInfo::Owning(owning_register_info) => match owning_register_info.extra {
+                Some(BitExtraInfo::Qubit { is_ancilla }) => is_ancilla,
+                _ => false,
+            },
+            RegisterInfo::Alias { extra, .. } => match extra {
+                Some(BitExtraInfo::Qubit { is_ancilla }) => *is_ancilla,
+                _ => false,
+            },
```

Comment:
Is this logic correct? I feel like you could have mixed `AncillaQubit` objects with regular `Qubit` in python without an issue since it was a subclass. So I'm not sure if this is correct since we're only inspecting the first referenced qubit object. I'm thinking either this needs to be `self.0.iter().any(...)`, self.0.iter().all(...)`, oralways return false for `Alias`. Either way maybe this warrants an upgrade release note? I don't think the exact behavior matching python is at all important here because this is the edge of an edge case, but more it is a potential difference we should document.

Reply:
We check the type safety more thoroughly in 883d82334.